source/server/connection_handler_impl.cc

```diff
@@ -293,54 +314,89 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
 
   const bool empty_filter_chain = !config_.filterChainFactory().createNetworkFilterChain(
-      *new_connection, filter_chain->networkFilterFactories());
+      *active_connection->connection_, filter_chain->networkFilterFactories());
   if (empty_filter_chain) {
-    ENVOY_CONN_LOG(debug, "closing connection: no filters", *new_connection);
-    new_connection->close(Network::ConnectionCloseType::NoFlush);
-    return;
+    ENVOY_CONN_LOG(debug, "closing connection: no filters", *active_connection->connection_);
+    active_connection->connection_->close(Network::ConnectionCloseType::NoFlush);
   }
 
-  onNewConnection(std::move(new_connection));
-}
-
-void ConnectionHandlerImpl::ActiveTcpListener::onNewConnection(
-    Network::ConnectionPtr&& new_connection) {
-  ENVOY_CONN_LOG(debug, "new connection", *new_connection);
-
   // If the connection is already closed, we can just let this connection immediately die.
-  if (new_connection->state() != Network::Connection::State::Closed) {
-    ActiveConnectionPtr active_connection(
-        new ActiveConnection(*this, std::move(new_connection), parent_.dispatcher_.timeSource()));
+  if (active_connection->connection_->state() != Network::Connection::State::Closed) {
+    ENVOY_CONN_LOG(debug, "new connection", *active_connection->connection_);
+    active_connection->connection_->addConnectionCallbacks(*active_connection);
     active_connection->moveIntoList(std::move(active_connection), connections_);
-    parent_.num_connections_++;
   }
 }
 
-ConnectionHandlerImpl::ActiveConnection::ActiveConnection(ActiveTcpListener& listener,
-                                                          Network::ConnectionPtr&& new_connection,
-                                                          TimeSource& time_source)
+namespace {
+// Structure used to allow a unique_ptr to be captured in a posted lambda. See below.
+struct RebalancedSocket {
+  Network::ConnectionSocketPtr socket;
+};
+using RebalancedSocketSharedPtr = std::shared_ptr<RebalancedSocket>;
+} // namespace
+
+void ConnectionHandlerImpl::ActiveTcpListener::post(Network::ConnectionSocketPtr&& socket) {
+  // It is not possible to capture a unique_ptr because the post() API copies the lambda, so we must
+  // bundle the socket inside a shared_ptr that can be captured.
+  // TODO(mattklein123): It may be possible to change the post() API such that the lambda is only
+  // moved, but this is non-trivial and needs investigation.
+  RebalancedSocketSharedPtr socket_to_rebalance = std::make_shared<RebalancedSocket>();
+  socket_to_rebalance->socket = std::move(socket);
+
+  parent_.dispatcher_.post([socket_to_rebalance, tag = config_.listenerTag(), &parent = parent_]() {
+    // TODO(mattklein123): We should probably use a hash table here to lookup the tag instead of
+    // iterating through the listener list.
+    for (const auto& listener : parent.listeners_) {
```

Comment:
Does this look up needed? I thought balancer should already picked the right listener to post as it's balancing across listeners with same ListenerConfig. No?

Reply:
Yeah it's needed. The listeners are per thread and we need to deal with the deletion case. This can be made more efficient per the comment but I think this is ok for now.

Reply:
Ah, right. Another approach I can think of is adding a bool in the listener as an indicator of whether this listener is deleted or not.

Reply:
Yeah, TBH I don't recall how we wound up with the listener tag concept. I would like to revisit this and I plan on coming back in here to do various cleanups including removing the dynamic_casts so I will take a look.